### PR TITLE
Vo/build link url

### PR DESCRIPTION
**Pull request by vanesa**
decorate line when document repo and file match Sentry extension settings

### CODE REVIEW

File: src/extension.ts [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+ * Verify if the params from the document URI match with the repo and file formats specified
+ * in the Sentry extension settings. If there is a match we know the document is enabled to send logs
+ * to Sentry and can send back the corresponding Sentry project ID.
+ * @param params params extracted from the document's URI.
+ * @param projects Sentry extension projects configurations.
+ * @return Sentry projectID this document reports to.
+ */
+export function matchSentryProject(params: Params, projects: Settings['sentry.projects']): SentryProject | undefined {
+    // Check if a Sentry project is associated with this document's repo and retrieve the project.
+    const project = projects!.find(p => !!new RegExp(p.patternProperties.repoMatch).exec(params.repo!))
```
Comment (lguychard) [inline src/extension.ts]: Please do not use the `!` operator liberally, if values are possibly `undefined`  or `null` you need to handle that case explicitly.

File: src/extension.ts [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+ * @returns repo and file part of URI.
+ */
+export function getParamsFromUriPath(textDocument: string): Params {
+    // TODO: Support more than just GitHub.
+    const repoPattern = /github\.com\/([^\?\#]+)/gi
+    const filePattern = /#([^\?\#\/]+)\/.*\.?$/gi
+    const repoM = repoPattern.exec(textDocument)
+    const fileM = filePattern.exec(textDocument)
+    return {
+        repo: repoM![1],
```
Comment (lguychard) [inline src/extension.ts]: Please do not use the `!` operator here. `exec()` may return null, and this code will throw a `TypeError` in that case.

File: src/extension.ts [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+interface SentryProject {
+    projectId: string
+    lineMatches: RegExp[]
+}
+
+const DECORATION_TYPE = sourcegraph.app.createDecorationType()
+const SETTINGSCONFIG = resolveSettings(sourcegraph.configuration.get<Settings>().value)
+
+export function activate(context: sourcegraph.ExtensionContext): void {
+    sourcegraph.workspace.onDidOpenTextDocument.subscribe(textDocument => {
```
Comment (lguychard) [inline src/extension.ts]: Why do you need this? You could just use `activeEditor`

File: src/extension.ts [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+ * @param errorQuery extracted from the error handling code matching the config matching pattern.
+ * @param sentryProjectId from the associated Sentry project receiving logs from the document's repo.
+ * @return URL to the Sentry unresolved issues stream page for this kind of query.
+ */
+function buildUrl(errorQuery: string, sentryProjectId: string): URL {
+    const url = new URL(
+        'https://sentry.io/organizations/' +
+            encodeURIComponent(SETTINGSCONFIG['sentry.organization']!) +
+            '/issues/?project=' +
+            encodeURIComponent(sentryProjectId) +
```
Comment (lguychard) [inline src/extension.ts]: [`URLSearchParams`](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams) would be helpful here.

File: src/extension.ts [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+        return
+    }
+
+    return {
+        projectId: project.projectId,
+        lineMatches: project.patternProperties.lineMatches,
+    }
+}
+
+function decorateEditor(editor: sourcegraph.CodeEditor, sentryProjectId: string, lineMatches: RegExp[]): void {
```
Comment (lguychard) [inline src/extension.ts]: I suggest refactoring this so that it calls a new function that returns `TextDocumentDecoration[]`, and adding tests for that new function (kind of like [`getBlameDecorations()`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph-git-extras@master/-/blob/src/blame.ts?diff=5f1e356db20824945ba6f50a803ff30d555cf785#L112-136))

File: src/extension.ts [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+                m = pattern.exec(line)
+                if (m) {
+                    decorations.push({
+                        range: new sourcegraph.Range(i, 0, i, 0),
+                        isWholeLine: true,
+                        after: {
+                            backgroundColor: '#e03e2f',
+                            color: 'rgba(255, 255, 255, 0.8)',
+                            contentText: ' View logs in Sentry » ',
+                            // Depending on the line matching pattern the query is indexed in position 1 or 2.
```
Comment (lguychard) [inline src/extension.ts]: For now this is true, but eventually you want to be able to specify which capture group should be used through configuration.

File: src/settings.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+ * See the "contributes.configuration" field in package.json for the canonical documentation on these properties.
+ */
+export interface Settings {
+    ['sentry.organization']?: string
+    ['sentry.projects']?: [
+        {
+            name: string
+            projectId: string
+            patternProperties: {
+                repoMatch: RegExp
```
Comment (lguychard) [inline src/settings.ts]: Just checking - you have yet to write the code that actually resolves the string values from the settings to `RegExp`, right?

File: src/extension.ts [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+const DECORATION_TYPE = sourcegraph.app.createDecorationType()
+const SETTINGSCONFIG = resolveSettings(sourcegraph.configuration.get<Settings>().value)
+
+export function activate(context: sourcegraph.ExtensionContext): void {
+    sourcegraph.workspace.onDidOpenTextDocument.subscribe(textDocument => {
+        const params: Params = getParamsFromUriPath(textDocument.uri)
+        const sentryProjects = SETTINGSCONFIG['sentry.projects']
+
+        // Retrieve the Sentry project that this document reports to.
+        const sentryProject = matchSentryProject(params, sentryProjects)
```
Comment (lguychard) [inline src/extension.ts]: This is business logic and should be moved outside of `activate()` and into a separate, testable function

File: src/settings.ts (hunk 7 above)
Comment (vanesa) [reply src/settings.ts]: No, I already do that in `settings`.

File: src/extension.ts [hunk 9]
```diff
@@ -11,107 +12,55 @@ interface Params {
...
-        return
-    }
-
-    // Check if document matches the file matching pattern specified under the Sentry extension configuration.
-    const fileMatched: boolean = !!project.patternProperties.fileMatches.find(
-        pattern => !!new RegExp(pattern).exec(params.file!)
-    )
-    if (!fileMatched) {
+function decorateEditor(editor: sourcegraph.CodeEditor, sentryProjectId?: string, lineMatches?: RegExp[]): void {
+    if (!editor) {
```
Comment (lguychard) [inline src/extension.ts]: `activeEditor` has this:

```typescript
filter((editor): editor is sourcegraph.CodeEditor => editor !== undefined)
```
And the `editor` argument here is typed as `editor: sourcegraph.CodeEditor`, not `editor: sourcegraph.CodeEditor | null`. You shouldn't need this check, have you run into cases where you did need it?

File: src/extension.ts [hunk 10]
```diff
@@ -121,20 +70,61 @@ function decorateEditor(editor: sourcegraph.CodeEditor, sentryProjectId: string,
...
+): sourcegraph.TextDocumentDecoration {
+    let decoration: sourcegraph.TextDocumentDecoration
+    decoration = {
+        range: new sourcegraph.Range(i, 0, i, 0),
+        isWholeLine: true,
+        after: {
+            backgroundColor: '#e03e2f',
+            color: 'rgba(255, 255, 255, 0.8)',
+            contentText: !SENTRYORGANIZATION
+                ? ' Add Sentry extension configurations to your settings to view logs in Sentry » '
```
Comment (lguychard) [inline src/extension.ts]: I like the idea of "hinting" decorations when the extension isn't configured, but the wording and the `»` here would make me think that these decorations are actionable, while they aren't. Two options:
- Remove `»`, add a more helpful hover message (this should only be a temporary fix)
- This link is actually active, and links to `$SOURCEGRAPH_URL/users/$USER/settings`. You could get the current username through graphQL.

Also: this text is quite long, which means it can end up being cropped quite easily. I suggest:

`Configure the Sentry extension to view logs` (33 chars shorter)

File: src/extension.ts [hunk 11]
```diff
@@ -11,107 +12,55 @@ interface Params {
...
-}
-
 const DECORATION_TYPE = sourcegraph.app.createDecorationType()
 const SETTINGSCONFIG = resolveSettings(sourcegraph.configuration.get<Settings>().value)
+const SENTRYORGANIZATION = SETTINGSCONFIG['sentry.organization']
+
+const COMMON_ERRORLOG_PATTERNS = [
+    /throw new Error+\(['"]([^'"]+)['"]\)/gi,
+    /console\.[^'"`]+\(['"`]([^'"`]+)['"`]\)/gi,
+    /log\.[^'"]+\(['"]([^'"]+)['"]\)/gi,
```
Comment (lguychard) [inline src/extension.ts]: Can you give a real-world example of this? It seems to match strings like `log."foo"`.

File: src/extension.ts [hunk 12]
```diff
@@ -11,107 +12,55 @@ interface Params {
...
-    lineMatches: RegExp[]
-}
-
 const DECORATION_TYPE = sourcegraph.app.createDecorationType()
 const SETTINGSCONFIG = resolveSettings(sourcegraph.configuration.get<Settings>().value)
+const SENTRYORGANIZATION = SETTINGSCONFIG['sentry.organization']
+
+const COMMON_ERRORLOG_PATTERNS = [
+    /throw new Error+\(['"]([^'"]+)['"]\)/gi,
+    /console\.[^'"`]+\(['"`]([^'"`]+)['"`]\)/gi,
```
Comment (lguychard) [inline src/extension.ts]: This appears to match strings like `console.("foo")`. Don't you want to match `/console\.(log|error|info|warn)/`?

File: src/extension.ts [hunk 13]
```diff
@@ -121,20 +70,61 @@ function decorateEditor(editor: sourcegraph.CodeEditor, sentryProjectId: string,
...
 /**
- * Build URL to the Sentry issues stream page with the Sentry Org, Sentry project ID, and query.
+ * Decorate a line that matches either the line match pattern from the Sentry extension configurations
+ * or that matches common error loggin patterns.
+ * @param i index for decoration range
+ * @param m line match containing error query
+ * @param sentryProjectId Sentry project id retrieved from Sentry extension settings
+ */
+export function decorateLine(
+    i: number,
```
Comment (lguychard) [inline src/extension.ts]: should be `index`, avoid single-letter argument names.

File: src/extension.ts [hunk 14]
```diff
@@ -121,20 +70,61 @@ function decorateEditor(editor: sourcegraph.CodeEditor, sentryProjectId: string,
...
+    m: RegExpExecArray,
```
Comment (lguychard) [inline src/extension.ts]: Should be `matches`, avoid single-letter argument names.

File: src/extension.ts [hunk 15]
```diff
@@ -121,20 +70,61 @@ function decorateEditor(editor: sourcegraph.CodeEditor, sentryProjectId: string,
...
+    sentryProjectId?: string
+): sourcegraph.TextDocumentDecoration {
+    let decoration: sourcegraph.TextDocumentDecoration
```
Comment (lguychard) [inline src/extension.ts]: You can declare this as `const` and assign it directly. Always declare things as `const` when possible, and avoid declaring variables as mutable unless absolutely necessary.

File: src/handler.ts [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+    if (!project) {
+        return
+    }
+
+    // Check if document matches the file matching pattern specified under the Sentry extension configuration.
+    if (project.patternProperties.fileMatches) {
+        if (!params.file) {
+            return
+        }
+        const fileMatched: boolean = project.patternProperties.fileMatches.some(
```
Comment (lguychard) [inline src/handler.ts]: No need for `: boolean`, the type will be inferred by the compiler. Always rely on inference when it's possible.

File: src/handler.ts [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+        return
+    }
+    // Check if a Sentry project is associated with this document's repo and retrieve the project.
+    const project = projects.find(p => !!new RegExp(p.patternProperties.repoMatch).exec(params.repo!))
+    if (!project) {
+        return
+    }
+
+    // Check if document matches the file matching pattern specified under the Sentry extension configuration.
+    if (project.patternProperties.fileMatches) {
```
Comment (lguychard) [inline src/handler.ts]: Remove the nested `if (!params.file)` and combine it with this statement.

File: src/handler.ts [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+ */
+export function matchSentryProject(
+    params: Params,
+    projects: Settings['sentry.projects'] | null
+): SentryProject | undefined {
+    if (!projects || !params.repo) {
+        return
+    }
+    // Check if a Sentry project is associated with this document's repo and retrieve the project.
+    const project = projects.find(p => !!new RegExp(p.patternProperties.repoMatch).exec(params.repo!))
```
Comment (lguychard) [inline src/handler.ts]: Calling `exec()` with a `null` `params.repo` will throw a TypeError at runtime. Handle the `null` case instead of using a non-null assertion `!` and add a test for this edge case to make sure it's handled.

File: src/handler.ts [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+        }
+        const fileMatched: boolean = project.patternProperties.fileMatches.some(
+            pattern => !!new RegExp(pattern).exec(params.file!)
+        )
+        if (!fileMatched) {
+            return
+        }
+    }
+
+    return {
```
Comment (lguychard) [inline src/handler.ts]: If `project.patternProperties.fileMatches` is falsy, this block will be reached. Is this intended (it doesn't seem like it). Add a unit test to handle this edge case.

File: src/handler.ts [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+ * check if the current document sends log events to Sentry and
+ * use these params build URL to the Sentry issues stream page.
+ * @param textDocument A document URI.
+ * @returns repo and file part of URI.
+ */
+export function getParamsFromUriPath(textDocument: string): Params {
+    // TODO: Support more than just GitHub.
+    const repoPattern = /github\.com\/([^\?\#\/]+\/[^\?\#\/]*)/gi
+    const filePattern = /#([^\?\#\/]+)\/.*\.?$/gi
+    const repoM = repoPattern.exec(textDocument)
```
Comment (lguychard) [inline src/handler.ts]: Naming: spell these out completely as `repoMatch`, `fileMatch`

File: src/extension.ts (hunk 11 above)
Comment (vanesa) [reply src/extension.ts]: It shouldn't match `log."foo"` but rather patterns like `log.Print("foo")` or `log.Printf("bar")`
image

File: src/extension.ts (hunk 12 above)
Comment (vanesa) [reply src/extension.ts]: I don't seem to be getting the same matches as you:
image

Comment (vanesa) [reply src/extension.ts]: But you are right: It will match wrong code like `console.l("test)` e.g. so maybe I should safeguard it with `(log|error|info|warn)`.

File: src/extension.ts (hunk 11 above)
Comment (vanesa) [reply src/extension.ts]: Ok, I'll limit it to the known `log.Print*()` options.

File: src/extension.ts (hunk 2 above)
Comment (vanesa) [reply src/extension.ts]: Yes, I want to work on a separate PR that will take care of all kinds of exceptions such as this one.

File: src/extension.ts (hunk 3 above)
Comment (vanesa) [reply src/extension.ts]: I'll refactor that part of the code in a separate PR.

File: src/extension.ts (hunk 10 above)
Comment (vanesa) [reply src/extension.ts]: Yes, this will be addressed in a separate PR.

File: src/extension.ts (hunk 5 above)
Comment (vanesa) [reply src/extension.ts]: Ok, will do in a separate PR.

File: src/extension.ts (hunk 8 above)
Comment (vanesa) [reply src/extension.ts]: Will do in next PR.

File: src/handler.ts (hunk 18 above)
Comment (vanesa) [reply src/handler.ts]: Will do in next PR.

File: src/handler.ts (hunk 19 above)
Comment (vanesa) [reply src/handler.ts]: Will do in next PR.

File: src/extension.ts (hunk 6 above)
Comment (vanesa) [reply src/extension.ts]: Added that as a TODO.

File: src/extension.ts (hunk 4 above)
Comment (vanesa) [reply src/extension.ts]: Will do this in the next PR.
